### PR TITLE
Improve plugin perf in large repos

### PR DESCRIPTION
**Pull request by markjm**
Hi @xz64, me again - I noticed some perf bottlenecks in very large project. In very large projects with thousands of modules, the IO operations really limit the performance of the plugin.

For some quick stats, in a project ending with ~300 identified modules, there were about 1M file IO calls. Most of them come from  pluginfilehandler's getModule running on duplicate files over and over. ModuleCache comes into picture too late and getModule's traversing upwards is very expensive.

Results on our large (unfortunately private so cant share specifics) as reported by SMP plugin - Before: 2min. After: 20sec

Summary of changes:
1. Upgraded typescript version because it was very easy and allowed me to use new typescript features -- we still transpile to es5, so no change in functionality or support.
2. Upgraded prettier along with that because old version doesnt support new syntax - the bulk of the changes are because of new prettier.
3. Dont require IO statSync check to determine if we found the module in the module directory, as the file-ness can be inferred based on the usage (and it doesnt seem to matter in some weird case where it is a dir anyway).
4. Use the parsed packageJson from the getModule, if available, instead of re-reading and parsing the package again in processModule. If its not available (as is the case when processModule is called from other locations), then we use old functionality to re-read it.
5. And the biggest perf change - include a simple cache object in PluginFileHandler to ensure we arent doing any work on duplicates. I also tested caching all io operations in WebpackFileSystem instead, but this proved a more fruitful approach. Caching in both places caused a lot of wasted memory.

### CODE REVIEW

File: src/PluginFileHandler.ts [hunk 1]
```diff
@@ -1,27 +1,40 @@
...
     if (filename === null || filename === undefined) {
       return null;
     }
 
     if (this.modulesDirectories !== null) {
       let foundInModuleDirectory = false;
       for (const modulesDirectory of this.modulesDirectories) {
-        if (this.fileSystem.isFileInDirectory(filename, modulesDirectory)) {
+        if (
+          this.fileSystem
```
Comment (markjm) [inline src/PluginFileHandler.ts:34]: See Summary of Changes pt 3

File: src/PluginFileHandler.ts [hunk 2]
```diff
@@ -67,6 +85,7 @@ class PluginFileHandler implements FileHandler {
     const packageJson: PackageJson = this.parsePackageJson(dirOfModule);
 
     return {
+      packageJson,
```
Comment (markjm) [inline src/PluginFileHandler.ts:88]: See Summary of Changes pt 4

File: src/PluginChunkReadHandler.ts [hunk 3]
```diff
@@ -45,25 +46,25 @@ class PluginChunkReadHandler implements WebpackChunkHandler {
...
+    module: Module | LicenseIdentifiedModule | null
   ) {
     if (module && !moduleCache.alreadySeenForChunk(chunk.name, module.name)) {
       const alreadyIncludedModule = moduleCache.getModule(module.name);
       if (alreadyIncludedModule !== null) {
         moduleCache.registerModule(chunk.name, alreadyIncludedModule);
       } else {
         // module not yet in cache
-        const packageJson: PackageJson = this.getPackageJson(module.directory);
+        const packageJson: PackageJson =
```
Comment (markjm) [inline src/PluginChunkReadHandler.ts:65]: See Summary of Changes pt 4

File: src/PluginFileHandler.ts [hunk 4]
```diff
@@ -1,27 +1,40 @@
...
     private buildRoot: string,
     private modulesDirectories: string[] | null,
-    private excludedPackageTest: ((packageName: string) => boolean)
+    private excludedPackageTest: (packageName: string) => boolean
   ) {}
 
   static PACKAGE_JSON: string = 'package.json';
 
-  getModule(filename: string): Module | null {
+  getModule(filename: string): LicenseIdentifiedModule | null {
```
Comment (markjm) [inline src/PluginFileHandler.ts:17]: See Summary of Changes pt 5

**Comment by markjm** [comment]
The prettier changes are a little cluttering, so I left comments where the actual changes are

**Comment by xz64** [comment]
These are some pretty major changes. Thank you for working on this! I will review it when I get a chance.

**Comment by xz64** [comment]
Published in v2.3.17. Thanks

**Pull request merged**